app/api/src/docker/openscad/openscad.js

```diff
@@ -12,25 +12,52 @@ const {
 
 const s3 = new AWS.S3()
 
+const openScadStlKey = (eventBody) => {
+  const { file } = JSON.parse(eventBody)
+  return `${makeHash(JSON.stringify(file))}.stl`
+}
+
 const preview = async (req, _context, callback) => {
   _context.callbackWaitsForEmptyEventLoop = false
   const eventBody = req.body
   console.log('eventBody', eventBody)
   const key = `${makeHash(eventBody)}.png`
+  const stlKey = openScadStlKey(eventBody)
+
   console.log('key', key)
 
+  const stlParams = {
+    Bucket: process.env.BUCKET,
+    Key: stlKey,
+  }
+
   const params = {
     Bucket: process.env.BUCKET,
     Key: key,
   }
-  const previousAsset = await checkIfAlreadyExists(params, s3)
+  const [previousAssetStl, previousAssetPng] = await Promise.all([
+    checkIfAlreadyExists(stlParams, s3),
+    checkIfAlreadyExists(params, s3),
+  ])
+  const type = previousAssetStl.isAlreadyInBucket ? 'stl' : 'png'
+  const previousAsset = previousAssetStl.isAlreadyInBucket
+    ? previousAssetStl
+    : previousAssetPng
   if (previousAsset.isAlreadyInBucket) {
     console.log('already in bucket')
     const response = {
       statusCode: 200,
       body: JSON.stringify({
-        url: getObjectUrl(params, s3),
-        consoleMessage: previousAsset.consoleMessage,
+        url: getObjectUrl(
+          {
+            Bucket: process.env.BUCKET,
+            Key: previousAssetStl.isAlreadyInBucket ? stlKey : key,
+          },
+          s3
+        ),
+        consoleMessage:
+          previousAsset.consoleMessage || previousAssetPng.consoleMessage,
```

Comment:
This is a bit messy. This is for the OpenSCAD preview, but now it check for both an image cache as well as an stl cache, if it finds either it will return one but with preference for the stl, so that if someone else has cached the 3d output by downloading it than users looking at the preview for the same shape will get the 3d view. It's only if they change the code it will fallback to images again.